Fixes #35601 - Search in navigation

MariaAga opened the pull request:
![Screenshot from 2023-05-08 19-07-09](https://user-images.githubusercontent.com/30431079/236886441-3e42b223-2972-48ca-b888-5dfd3dc06e8c.png)

MariSvirik commented:
@MariaAga could you pls add also a screenshot of an inactive search bar (default state)?

1. When you make the search bar active by clicking on it (but not typing anything) -  is it going to give you all the suggestions? Or would it wait for any letters?

2. Do you have a screenshot of an empty state?

3. What is the font size?

MariaAga commented:
Here is a link to a demo which hopefully answer most questions: https://hmj7ff.csb.app/ (its a bit jumpy when navigating)
Font size for the result item is: `font-size 16px, font-weight 400`
Font size for the result item path is: `font-size 12px, font-weight 400`

MariSvirik commented:
**1.** I've noticed that sometimes words in the gray breadcrumb are split in the middle e.g. Recurring logic or Config management. Can we fix that somehow? I know it probably depends on the size of your screen. But could we have a rule not to split the words in the middle and just move them to the next line?

<img width="217" alt="Screenshot 2023-07-05 at 15 56 26" src="https://github.com/theforeman/foreman/assets/51960547/0c67e19c-2274-44fb-b50d-be65e009c005">


**2.** Also, not sure if that's a bug or a feature:
I need to click twice to actually be able to type anything. So there are actually three states of search which I think is unnecessary. 

<img width="269" alt="Screenshot 2023-07-05 at 16 03 27" src="https://github.com/theforeman/foreman/assets/51960547/2e68324d-87ab-4ee6-815f-be6c2e795fa4">
<img width="285" alt="Screenshot 2023-07-05 at 16 03 31" src="https://github.com/theforeman/foreman/assets/51960547/9c78db09-2e4e-4523-b969-d69bbfd8ed77">
<img width="267" alt="Screenshot 2023-07-05 at 16 03 44" src="https://github.com/theforeman/foreman/assets/51960547/b6a3d3bd-9e88-4a8d-af7e-ef295bdc3be5">



**3.** What's more, there is an error when I start typing.

<img width="1230" alt="Screenshot 2023-07-05 at 16 06 06" src="https://github.com/theforeman/foreman/assets/51960547/55c34619-aea0-41c3-aa5b-b21c60ae22e9">

MariaAga commented:
@MariSvirik Thanks!
1. Fixed in the demo https://hmj7ff.csb.app/host_statuses
2. I'm using the expendable search like here: https://www.patternfly.org/v4/components/search-input#with-expandable-button
Which does focus on the text if users click on the search glass, originally the space next to the search glass icon shouldnt be clickable at all.
Should I just use a normal search instead of an expandable one?
3. Fixed in the demo

adamruzicka commented:
@MariSvirik ping

MariaAga commented:
@MariSvirik  I also made a demo that doesnt use expendable search, but just a search input here: https://l9hdpm.csb.app/audits
What do you think?

MariaAga committed:
Fixes #35601 - Search in navigation

MariaAga commented:
Talked to Maria S and we decided on this design:
![Screenshot from 2023-10-17 21-45-54](https://github.com/theforeman/foreman/assets/30431079/a6a318b5-6dea-4eff-b800-17d6ed9afd6c)
@ofedoren Can you please review this PR?

ofedoren commented on webpack/assets/javascripts/react_app/components/Layout/NavigationSearch.js:
```diff
@@ -0,0 +1,217 @@
+      }
+    };
+    const handleMenuKeys = event => {
+      // If there is a hint while the browser focus is on the search input, tab or right arrow will 'accept' the hint value
```
That doesn't seem to be working... Also, I don't see where is `setHint` being used except reseting below.

MariaAga replied:
leftovers from copying PF logic

ofedoren commented on webpack/assets/javascripts/react_app/components/Layout/NavigationSearch.js:
```diff
@@ -0,0 +1,217 @@
+    setIsAutocompleteOpen(false);
+  };
+
+  React.useEffect(() => {
```
nit: why is here explicit `React.useEffect`, but `useState` and `useRef` are imported?

MariaAga replied:
nope, fixed to match the rest

ofedoren reviewed (commented):
Thanks, @MariaAga! Great addition. No more trying to remember where is $ITEM located :)

Few comments:
 - Is there a way to trigger the nav search via keyboard? Looking at the tickets, there was an experimental key **`** as an example.
  - It would be nice if the first item from the suggestion list was auto-selected, so I don't need to use arrows for explicit selection.
   - Some comments inline.
   - Test failures seem to be related?
   - Huge thanks for the explanations in the comments :)

MariaAga commented:
Thanks! 

> Is there a way to trigger the nav search via keyboard? Looking at the tickets, there was an experimental key ` as an example.

added CTRL + SHIFT + F shortcut as ` was mentioned not being easily accessible in some keyboard layouts

> It would be nice if the first item from the suggestion list was auto-selected, so I don't need to use arrows for explicit selection.

I'm not sure if that what you meant, but now if a user starts typing in the search and clicks ENTER the page will redirect to the first result 

Added more specific selectors to failing tests

ofedoren commented on webpack/assets/javascripts/react_app/components/Layout/NavigationSearch.js:
```diff
@@ -0,0 +1,231 @@
+  const searchInput = (
+    <SearchInput
+      value={value}
+      placeholder={__('Search and go')}
```
Could we change it to `Ctrl + Shift + F to search`? I mean, how would one know what to use?

Not a big deal though, since that combination is used in a lot of editors...

MariaAga replied:
I talked to Maria from UX about this, and since we (and other projects) dont show the `/` shortcut in the search bar placeholder we shouldn't do it here as well. I'll get it in the docs and I added it to the about page 
![image](https://github.com/theforeman/foreman/assets/30431079/2151bdf4-e4c8-498d-aad6-9d07a37bc923)

ofedoren reviewed (dismissed):
Thanks, @MariaAga! One inline question, but otherwise good to go :)

ofedoren reviewed (approved):
Love it, thanks @MariaAga! Not sure though if we can merge it now or should we wait for branching...

ofedoren reviewed (approved):
Thanks, @MariaAga, apparently with this one we can go further :) Promise me, you'll re-check we didn't break anything for most wanted plugins :D
